content/blog/en/dojo-resource-widgets.md

```diff
@@ -0,0 +1,465 @@
+            />
+          ))}
+    </section>
+  );
+});
+
+export default HackerNewList;
+```
+
+Ok, there's a lot going on here, so let's break it down.
+
+First thing we do is import the modules to create the widget and use some middleware.
+
+```ts
+import { create, tsx } from "@dojo/framework/core/vdom";
+import { createResourceMiddleware } from "@dojo/framework/core/middleware/resources";
+import { createICacheMiddleware } from "@dojo/framework/core/middleware/icache";
+
+import HackerNewItem from "./HackerNewItem";
+import hnTemplate from "../hnTemplate";
+
+interface ListState {
+  search: string;
+  readItemIds: string[];
+  page: number;
+}
+```
+
+The `createResourceMiddleware` let's you create a typed resource interface to work with resource templates.
+The `createICacheMiddleware` allows you to create a typed `icache` middleware. It's more type friendly than simply using `icache` in your widgets. Developer ergonomics people. Someone may have to work on this widget six months down the road and _better types, no tears to wipe_.
+
+We also import a simple widget to display each news item and our template. `ListState` is the interface for our widgets local state. Makes it easier to track what page we are on and which news items we're filtering out.
+
+Now we can create the middleware and our widget factory method.
+
+```tsx
+const resource = createResourceMiddleware();
+const icache = createICacheMiddleware<ListState>();
+const factory = create({ icache, resource });
+```
+
+This pretty straight forward, the factory method will let us create widgets with local state and a resource API. Now we can move on to the widget, which will cover in sections.
+
+```tsx
+const HackerNewList = factory(function HackerNewList({
+  id,
+  middleware: { icache, resource }
+}) {
+  // de-structure resource API controls
+  const { createOptions, getOrRead, isFailed, isLoading } = resource;
+  // create an set of options to work with, with the widget id
+  const options = createOptions(id);
+  // get the current search term
+  const search = icache.getOrSet("search", "TypeScript");
+  const page = icache.getOrSet("page", 1);
+  const currentOptions = options({ size: 20, page, query: { search } });
+  // resource aware widgets let you know if they are loading or there was an error
+  // results are return in arrays of pages (can request multiple pages)
+  const [result] = getOrRead(hnTemplate, currentOptions);
+  const loading = isLoading(hnTemplate, currentOptions);
+  const failed = isFailed(hnTemplate, currentOptions);
+  const readItemIds = icache.getOrSet("readItemIds", []);
+  ...
+});
+```
+
+The `resource` middleware provides an API to interact with the resource template.
+
+* `createOptions`: This is used to create a payload specific to your widget via `createOptions(id)`, where `id` is the id of the widget. This is useful, because you can have multiple widgets using the same resource in your application, so internally, Dojo will manage that.
```

Comment:
```suggestion
* `createOptions`: Create a payload specific to your widget via `createOptions(id)`, where `id` is the widget's id. This method is useful because you can have multiple widgets using the same resource in your application, so internally Dojo will manage that automatically.
```